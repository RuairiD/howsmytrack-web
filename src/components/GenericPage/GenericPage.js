import React from 'react';

import apiRoot from '../../apiRoot';

import { Divider, Layout, PageHeader, Typography } from 'antd';
import MenuBar from '../MenuBar/MenuBar';

const REFRESH_TOKEN_FROM_COOKIE_MUTATION = `mutation RefreshTokenFromCookie {
    refreshTokenFromCookie {
        token
    }
}`;

type Props = {
    title: string,
    subTitle: string,
    hideMenu: boolean,
    isMobile: boolean,
};

// TODO: I hate this but it was a quick and dirty way of making sure the page filled the container and
// stayed the same width as other pages which had text they actually needed.
const TRANSPARENT_TEXT = (
    <Typography.Paragraph style={{ opacity: 0, height: 0 }}>
        Lorem ipsum dolor sit amet, consectetur adipiscing elit. Praesent sed hendrerit leo. Donec vitae risus et ante egestas sollicitudin at a mi. Duis fringilla a mi ut congue. Sed elit nunc, mollis sit amet interdum id, viverra vitae ligula. Proin eu risus vitae turpis fermentum maximus. Phasellus finibus enim nibh, non cursus lorem auctor interdum. Sed a ex id magna ultricies gravida. Proin sit amet sem at quam tristique tristique sit amet sed augue. Cras in bibendum risus, eu consequat turpis. In sed commodo augue. Donec nibh nulla, viverra quis mi tincidunt, pulvinar interdum neque. 
    </Typography.Paragraph>
);

class GenericPage extends React.Component<Props> {
    /*
     * Component for displaying generic page with children.
     * Also responsible for refreshing JWT token on pageload.
     */

    componentDidMount() {
        fetch(apiRoot +'/graphql/', {
            method: 'POST',
            headers: {
                'Content-Type': 'application/json',
                'Accept': 'application/json',
            },
            body: JSON.stringify({
                query: REFRESH_TOKEN_FROM_COOKIE_MUTATION,
            }),
            credentials: 'include',
        });
    };

    getSiderWidth = () => {
        // Ideally, 560px+ of content should be visible besides the Sider.
        // A Sider of <200px is awkward though and similarly, a Sider of >256px
        // is redundant.
        let siderWidth = window.screen.width - 560;
        if (siderWidth < 200) {
            siderWidth = 200;
        } else if (siderWidth > 256) {
            siderWidth = 256;
        }
        return siderWidth + 'px';
    };

    render() {
        return (
            <Layout className="page-container">
                <div className="page-container-inner">
                    {!this.props.hideMenu && this.props.isMobile && <MenuBar isMobile={this.props.isMobile} />}
                    <Layout.Content>
                        <Layout>
                            {!this.props.hideMenu && !this.props.isMobile && <Layout.Sider theme="light" width={this.getSiderWidth()}>
                                <MenuBar />
                            </Layout.Sider>}
                            <Layout.Content className="page-content">
                                {!this.props.hideMenu && <PageHeader
                                    ghost={false}
                                    title={this.props.title}
                                    subTitle={this.props.subTitle}
                                />}
                                <Layout className="content">
                                    <Layout.Content>
                                        {this.props.children}
                                        {TRANSPARENT_TEXT}
                                    </Layout.Content>

                                    <Layout.Footer style={{ textAlign: 'center' }}>
                                        <Divider />
                                        <Typography.Paragraph>
                                            <Typography.Text strong>
                                                <a target="_blank" rel="noopener noreferrer" href="http://ruairidorrity.com">ruairi dorrity</a> &#47;&#47; <a target="_blank" rel="noopener noreferrer" href="http://ruairidx.com">ruairi dx</a>
                                            </Typography.Text>
                                        </Typography.Paragraph>
                                        <Typography.Paragraph>
                                            <Typography.Text strong>source</Typography.Text><br />
                                            <Typography.Text>
                                                <a target="_blank" rel="noopener noreferrer" href="https://github.com/ruairid/howsmytrack-api">api</a> &#47;&#47; <a target="_blank" rel="noopener noreferrer" href="https://github.com/ruairid/howsmytrack-web">web</a>
                                            </Typography.Text>
                                        </Typography.Paragraph>
                                    </Layout.Footer>
                                </Layout>
                            </Layout.Content>
                        </Layout>
                    </Layout.Content>
                </div>
            </Layout>
        );
    }
}

export default GenericPage;
