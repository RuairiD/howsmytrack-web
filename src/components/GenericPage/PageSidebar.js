import React from "react";

import { Layout } from "antd";
import MenuBar from "../MenuBar/MenuBar";

const getSiderWidth = () => {
    // Ideally, 560px+ of content should be visible besides the Sider.
    // A Sider of <200px is awkward though and similarly, a Sider of >256px
    // is redundant.
    let siderWidth = window.screen.width - 560;
    if (siderWidth < 200) {
        siderWidth = 200;
    } else if (siderWidth > 256) {
        siderWidth = 256;
    }
    return `${siderWidth}px`;
};

const PageSidebar = () => (
    <Layout.Sider theme="light" width={getSiderWidth()}>
        <MenuBar />
    </Layout.Sider>
);

export default PageSidebar;
