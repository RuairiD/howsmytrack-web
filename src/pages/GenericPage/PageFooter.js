import React from "react";

import { Divider, Typography } from "antd";
import { A, Div } from "lemon-reset";

const PageFooter = () => (
    <Div style={{ textAlign: "center" }}>
        <Divider />
        <Typography.Paragraph>
            <Typography.Text strong>
                <A target="_blank" rel="noopener noreferrer" href="http://ruairidorrity.com">ruairi dorrity</A> &#47;&#47; <A target="_blank" rel="noopener noreferrer" href="http://ruairidx.com">ruairi dx</A>
            </Typography.Text>
        </Typography.Paragraph>
        <Typography.Paragraph>
            <Typography.Text strong>source</Typography.Text><br />
            <Typography.Text>
                <A target="_blank" rel="noopener noreferrer" href="https://github.com/ruairid/howsmytrack-api">api</A> &#47;&#47; <A target="_blank" rel="noopener noreferrer" href="https://github.com/ruairid/howsmytrack-web">web</A>
            </Typography.Text>
        </Typography.Paragraph>
    </Div>
);

export default PageFooter;
