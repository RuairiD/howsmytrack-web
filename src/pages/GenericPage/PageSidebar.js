import React from "react";

import { Affix, Layout } from "antd";
import { Div } from "lemon-reset";
import MenuBarController from "../../components/MenuBar/MenuBarController";

const getSiderWidth = () => {
    // Ideally, 560px+ of content should be visible besides the Sider.
    // A Sider of <200px is awkward though and similarly, a Sider of >256px
    // is redundant.
    // TODO: not truly responsive
    let siderWidth = window.screen.width - 560;
    if (siderWidth < 200) {
        siderWidth = 200;
    } else if (siderWidth > 256) {
        siderWidth = 256;
    }
    return `${siderWidth}px`;
};

const PageSidebar = () => (
    <Layout.Sider theme="light" width={getSiderWidth()}>
        <Affix>
            <Div>
                <MenuBarController />
            </Div>
        </Affix>
    </Layout.Sider>
);

export default PageSidebar;
