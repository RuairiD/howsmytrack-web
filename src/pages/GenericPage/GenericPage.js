import React, { useEffect } from "react";

import { Layout, PageHeader, Typography } from "antd";
import { Div } from "lemon-reset";
import axios from "axios";
import apiRoot from "../../apiRoot";
import MenuBar from "../../components/MenuBar/MenuBar";
import PageFooter from "./PageFooter";
import PageSidebar from "./PageSidebar";

// TODO: UserSettingsPage looks weird because it shrinks to fit the small amount of content.
// Fix the CSS and remove this when you have time.
const TransparentText = () => (
    <Typography.Paragraph style={{ opacity: 0, height: 0, width: "100%", userSelect: "none", overflowY: "hidden" }}>
        Lorem ipsum dolor sit amet, consectetur adipiscing elit. Praesent sed hendrerit leo. Donec vitae risus et ante egestas sollicitudin at a mi. Duis fringilla a mi ut congue. Sed elit nunc, mollis sit amet interdum id, viverra vitae ligula. Proin eu risus vitae turpis fermentum maximus. Phasellus finibus enim nibh, non cursus lorem auctor interdum. Sed a ex id magna ultricies gravida. Proin sit amet sem at quam tristique tristique sit amet sed augue. Cras in bibendum risus, eu consequat turpis. In sed commodo augue. Donec nibh nulla, viverra quis mi tincidunt, pulvinar interdum neque.
    </Typography.Paragraph>
);

const REFRESH_TOKEN_MUTATION = `mutation RefreshToken {
    refreshToken {
        payload
    }
}`;

type Props = {
    title: string,
    subTitle: string,
    hideMenu: boolean,
};

const GenericPage = ({ title, subTitle, hideMenu, children }: Props) => {
    /*
     * Component for displaying generic page with children.
     * Also responsible for refreshing JWT token on pageload.
     */
    useEffect(() => {
        axios.post(`${apiRoot}/graphql/`, {
            query: REFRESH_TOKEN_MUTATION,
        });
    }, []);

    return (
        <Layout className="page-container">
            <Div className="page-container-inner">
                {!hideMenu && (
                    <Div className="d-flex d-md-none">
                        <MenuBar isMobile />
                    </Div>
                )}
                <Layout.Content>
                    <Layout>
                        {!hideMenu && (
                            <Div className="d-none d-md-flex">
                                <PageSidebar />
                            </Div>
                        )}
                        <Layout.Content
                            className="page-content"
                            style={{
                                // Fixes weird bug where multiple containers on the page
                                // scroll at once, causing erratic scrolling behaviour.
                                overflowX: "visible",
                            }}
                        >
                            {!hideMenu && (
                                <PageHeader
                                    ghost={false}
                                    title={title}
                                    subTitle={subTitle}
                                />
                            )}
                            <Layout className="content">
                                <Layout.Content>
                                    {children}
                                </Layout.Content>
                                <Layout.Footer>
                                    <PageFooter />
                                    <TransparentText />
                                </Layout.Footer>
                            </Layout>
                        </Layout.Content>
                    </Layout>
                </Layout.Content>
            </Div>
        </Layout>
    );
};

export default GenericPage;
