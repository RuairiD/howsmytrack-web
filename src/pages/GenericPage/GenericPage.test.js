import React from "react";
import { mount, shallow } from "enzyme";
import axios from "axios";
import { Div } from "lemon-reset";
import waitForExpect from "wait-for-expect";
import toJson from "enzyme-to-json";
import GenericPage from "./GenericPage";

jest.mock("axios");
jest.mock("react-redux");

describe("GenericPage", () => {
    beforeEach(() => {
        axios.post.mockClear();
    });

    it("makes a request to refresh the JWT token on pageload", async () => {
        // Sanity check to ensure mock cleared appropriately
        expect(axios.post.mock.calls.length).toBe(0);

        const wrapper = mount(
            <GenericPage
                title="title"
                subTitle="subTitle"
            >
                <Div className="test-children" />
            </GenericPage>,
        );

        await waitForExpect(() => {
            wrapper.update();
            expect(axios.post.mock.calls.length).toBeGreaterThan(0);
        });
    });

    it("renders a sidebar menu and a collapsible menu with desktop/mobile utility classes", () => {
        const wrapper = shallow(
            <GenericPage
                title="title"
                subTitle="subTitle"
            >
                <Div className="test-children" />
            </GenericPage>,
        );

        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it("renders no menu or PageHeader when hideMenu is true", () => {
        const wrapper = shallow(
            <GenericPage
                title="title"
                subTitle="subTitle"
                hideMenu
            >
                <Div className="test-children" />
            </GenericPage>,
        );

        expect(toJson(wrapper)).toMatchSnapshot();
    });
});
