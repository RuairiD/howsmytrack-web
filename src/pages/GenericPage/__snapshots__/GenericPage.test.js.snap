// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GenericPage renders a sidebar menu and a collapsible menu with desktop/mobile utility classes 1`] = `
<Layout
  className="page-container"
>
  <Div
    className="page-container-inner"
  >
    <Div
      className="d-flex d-md-none"
    >
      <MenuBar
        isMobile={true}
      />
    </Div>
    <Content>
      <Layout>
        <Div
          className="d-none d-md-flex"
        >
          <PageSidebar />
        </Div>
        <Content
          className="page-content"
          style={
            Object {
              "overflowX": "visible",
            }
          }
        >
          <PageHeader
            backIcon={
              <Icon
                type="arrow-left"
              />
            }
            ghost={false}
            subTitle="subTitle"
            title="title"
          />
          <Layout
            className="content"
          >
            <Content>
              <Div
                className="test-children"
              />
            </Content>
            <Footer>
              <PageFooter />
              <TransparentText />
            </Footer>
          </Layout>
        </Content>
      </Layout>
    </Content>
  </Div>
</Layout>
`;

exports[`GenericPage renders no menu or PageHeader when hideMenu is true 1`] = `
<Layout
  className="page-container"
>
  <Div
    className="page-container-inner"
  >
    <Content>
      <Layout>
        <Content
          className="page-content"
          style={
            Object {
              "overflowX": "visible",
            }
          }
        >
          <Layout
            className="content"
          >
            <Content>
              <Div
                className="test-children"
              />
            </Content>
            <Footer>
              <PageFooter />
              <TransparentText />
            </Footer>
          </Layout>
        </Content>
      </Layout>
    </Content>
  </Div>
</Layout>
`;
